sql/hive/src/test/scala/org/apache/spark/sql/hive/test/TestHive.scala

```diff
@@ -647,3 +648,18 @@ private[sql] class TestHiveSessionStateBuilder(
 
   override protected def newBuilder: NewBuilder = new TestHiveSessionStateBuilder(_, _)
 }
+
+private[hive] object HiveTestJars {
+  def getHiveContribJar: File = getJarFile("org.apache.hive:hive-contrib")
+  def getHiveHcatalogCoreJar: File = getJarFile("org.apache.hive.hcatalog:hive-hcatalog-core")
+
+  private def getJarFile(coordinate: String): File = {
+    // isTransitive = false: Only direct dependencies should be resolved
+    val filePath = SparkSubmitUtils.resolveMavenCoordinates(
+      s"$coordinate:${HiveUtils.builtinHiveVersion}",
+      SparkSubmitUtils.buildIvySettings(
+        Some("https://repository.apache.org/content/repositories/releases"), None),
+      isTransitive = false)
+    new File(filePath)
+  }
```

Comment:
cc @srowen @dongjoon-hyun @HyukjinKwon Do you think this is stable?

Reply:
I don't know how stable this way it is. While Sean or Dongjoon might have an idea about this, I would verify if `--jars` in that condition is still tested, and run the tests multiple times to show practically this way works at least.

Reply:
I think this could be OK. I was really thinking of refactoring the logic that directly downloads the JAR, found in `HiveExternalCatalogVersionsSuite.tryDownloadSpark`. However in retrospect, I think that logic is a little bit tied to downloading a release tarball, not a JAR. 

If you just want single JARs, how about just constructing the URL for it on `repository.apache.org` and download? I don't think you have to resolve it?

Reply:
Do you prefer this?
```scala
  private val repository = "https://repository.apache.org/content/repositories/releases/"
  private val hiveContribUrl = s"${repository}org/apache/hive/hive-contrib/" +
    s"${HiveUtils.builtinHiveVersion}/"
  private val hiveHcatalogCoreUrl = s"${repository}org/apache/hive/hcatalog/hive-hcatalog-core/" +
    s"${HiveUtils.builtinHiveVersion}/"

  def getHiveContribJar: File = getFileFromUrl(hiveContribUrl,
    s"hive-contrib-${HiveUtils.builtinHiveVersion}.jar")
  def getHiveHcatalogCoreJar: File = getFileFromUrl(hiveHcatalogCoreUrl,
    s"hive-hcatalog-core-${HiveUtils.builtinHiveVersion}.jar")

  private def getFileFromUrl(urlString: String, filename: String): File = {
    val hiveTestJars = new File("/tmp/test-spark/hiveTestJars")
    if (!hiveTestJars.exists()) {
      hiveTestJars.mkdirs()
    }
    val targetFile = new File(hiveTestJars, filename)
    if (!targetFile.exists() || !(targetFile.length() > 0)) {
      val conf = new SparkConf
      val securityManager = new org.apache.spark.SecurityManager(conf)
      val hadoopConf = new Configuration

      // propagate exceptions up to the caller of getFileFromUrl
      Utils.doFetchFile(urlString, hiveTestJars, filename, conf, securityManager, hadoopConf)
    }
    targetFile
  }
```

Reply:
Yes, that kind of thing. It could be a little simpler (e.g. no need for defs for the `getHiveContribJar` et al, and I think you can pass a null conf/securityManager to download an HTTPS URL) but this seems simpler than trying to resolve the artifact.

Reply:
Done

Reply:
```scala
  private def getJarFromUrl(urlString: String): File = {
    val fileName = urlString.split("/").last
    val targetFile = new File(hiveTestJarsDir, fileName)
    if (!targetFile.exists()) {
      Utils.doFetchFile(urlString, hiveTestJarsDir, fileName, new SparkConf, null, null)
    }
    targetFile
  }
```
The method I think it may be incorrect. Why not change to:
```scala
  private def getJarFromUrl(urlString: String): File = {
    val fileName = urlString.split("/").last
    Utils.doFetchFile(urlString, hiveTestJarsDir, fileName, new SparkConf, null, null)
  }
```
Because the `targetFile` and the return value of `Utils.doFetchFile` may be different.
In my testing:
`targetFile`:
/private/var/folders/pl/0j98k60s5zs7lz4kthvkn2dm3954vp/T/spark-c8c1a791-bef2-4e36-bd99-e03d89fd7d27/hive-contrib-1.2.1.jar
`Utils.doFetchFile`:
/private/var/folders/pl/0j98k60s5zs7lz4kthvkn2dm3954vp/T/spark-fdbe0ba4-a929-499a-9a29-55644bcd31b0/hive-contrib-1.2.1.jar

Reply:
Thank you @LantaoJin My change is to avoid downloading these two files many times.